AMQ-9392 - Rework fix for read check Timer leak

cshannon opened the pull request:
This commit updates the previous fix for preventing a read check Timer
leak by the inactivity monitor when TCP connections fail
This is a follow on to https://github.com/apache/activemq/pull/1119

This updated version has the following improvements:

  1) The logic to cancel timers and shutdown the executor has been
consolidated into one location and only done during stopping of the
transport. Previously, the update could stop the task/executor during
normal broker operation (see next point)
  2) stopConnectCheckTask() is called under normal circumstances during
initialization of a connection and the previous change caused the read
task to be cancelled and the executor to be stopped, only to be
immediately restarted again during the first connection creation.
  3) Null checks were added to guarantee no null pointer issues when
referencing tasks/timers during stop
  4) Helper methods and some comments were added to simplify the code
and make it easier to understand what is going on.

asanguinetti committed:
[AMQ-9392] Prevent InactivityMonitor read check Timer leak when TCP connection fails (#1119)

* Adding test.

* Naive fix.

* Fixing OpenWireConnectionTimeoutTest.
- It seems checking configuredOk can't be done that early.
- Falling back to let the Timer be created and make sure it is disposed of.

cshannon committed:
AMQ-9392 - Rework fix for read check Timer leak

This commit updates the previous fix for preventing a read check Timer
leak by the inactivity monitor when TCP connections fail
This is a follow on to #1119

This updated version has the following improvements:

  1) The logic to cancel timers and shutdown the executor has been
consolidated into one location and only done during stopping of the
transport. Previously, the update could stop the task/executor during
normal broker operation (see next point)
  2) stopConnectCheckTask() is called under normal circumstances during
initialization of a connection and the previous change caused the read
task to be cancelled and the executor to be stopped, only to be
immediately restarted again during the first connection creation.
  3) Null checks were added to guarantee no null pointer issues when
referencing tasks/timers during stop
  4) Helper methods and some comments were added to simplify the code
and make it easier to understand what is going on.

jbonofre reviewed (approved):
Much cleaner approach. Thanks !

jbonofre commented:
I'm just waiting Jenkins before merging (it worked locally on my machine).

cshannon committed:
AMQ-9392 - Add extra inactivty monitor test case

This test randomly creates 250 different types of connections to verify
the read/write tasks are cancelled after all the connections have been
cleaned up

mattrpav reviewed (commented):
LGTM
